core/tee/tee_fs_rpc.c

```diff
@@ -0,0 +1,600 @@
+/*
+ * Copyright (c) 2016, Linaro Limited
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <assert.h>
+#include <kernel/thread.h>
+#include <mm/core_memprot.h>
+#include <stdlib.h>
+#include <string.h>
+#include <tee/tee_fs.h>
+#include <tee/tee_fs_rpc.h>
+#include <trace.h>
+#include <util.h>
+
+#define RPC_FAILED -1
+
+struct tee_fs_dir {
+	int nw_dir;
+	struct tee_fs_dirent d;
+};
+
+static TEE_Result tee_fs_rpc_send_cmd(int cmd_id, struct tee_fs_rpc *bf_cmd,
+				      void *data, uint32_t len, uint32_t mode)
+{
+	TEE_Result ret;
+	struct optee_msg_param params;
+	paddr_t phpayload = 0;
+	uint64_t cpayload = 0;
+	struct tee_fs_rpc *bf;
+	int res = TEE_ERROR_GENERIC;
+
+	assert(cmd_id == OPTEE_MSG_RPC_CMD_FS ||
+	       cmd_id == OPTEE_MSG_RPC_CMD_SQL_FS);
+
+	thread_rpc_alloc_payload(sizeof(struct tee_fs_rpc) + len,
+				 &phpayload, &cpayload);
+	if (!phpayload)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	if (!ALIGNMENT_IS_OK(phpayload, struct tee_fs_rpc))
+		goto exit;
+
+	bf = phys_to_virt(phpayload, MEM_AREA_NSEC_SHM);
```

Comment:
i didn't see `thread_rpc_alloc_payload()` checking that allocated buffers really are inside the SHM.
if return buffer is physically at end of SHM, accessing it could overflow to some secure (or not) memory.